crates/astria-sequencer/src/mempool/recently_included_transactions.rs

```diff
@@ -0,0 +1,424 @@
+use tendermint::abci::types::ExecTxResult;
+use tokio::time::Instant;
+
+/// The maximum number of transactions that can be stored in the [`RecentlyIncludedTransactions`]
+/// cache. This number is calculated based on an assumed timeout time of 1 minute and a maximum
+/// throughput of 1000 transactions per second.
+const MAX_RECENTLY_INCLUDED_TRANSACTIONS: usize = 60_000;
+/// The timeout time for the transactions in the [`RecentlyIncludedTransactions`] cache.
+const MAX_TIME_RECENTLY_INCLUDED_TRANSACTIONS_SECS: u64 = 60;
+
+pub(super) struct IncludedTransactionMetadata {
+    height: u64,
+    result: ExecTxResult,
+}
+
+impl IncludedTransactionMetadata {
+    pub(super) fn height(&self) -> u64 {
+        self.height
+    }
+
+    pub(super) fn result(&self) -> &ExecTxResult {
+        &self.result
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, thiserror::Error)]
+pub(super) enum IncludedTransactionError {
+    #[error("Transaction ID {0} already exists in the cache")]
+    IdAlreadyExists(TransactionId),
+    #[error("Transaction insertion failed due to an internal error")]
+    InsertionInternalError,
+}
+
+/// A cache of transaction data for transactions which were recently included in a block. The cache
+/// has a maximum size of [`MAX_RECENTLY_INCLUDED_TRANSACTIONS`] transactions. Transactions are
+/// cleared from the cache after [`MAX_TIME_RECENTLY_INCLUDED_TRANSACTIONS_SECS`] seconds if the
+/// cache is not full.
+///
+/// Supports O(1) lookups by transaction ID while maintaining O(1) insertion and removal of oldest
+/// element. Worst case time complexity for cleaning stale transactions is O(n), where n is the
+/// number of transactions in the cache.
+pub(super) struct RecentlyIncludedTransactions {
+    /// Transaction IDs in chronological order (oldest first)
+    timestamped_ids: VecDeque<(TransactionId, Instant)>,
+    /// Hash map containing transaction metadata, keyed by transaction ID
+    transactions: HashMap<TransactionId, IncludedTransactionMetadata>,
+}
+
+impl RecentlyIncludedTransactions {
+    pub(super) fn new() -> Self {
+        Self {
+            timestamped_ids: VecDeque::new(),
+            transactions: HashMap::new(),
+        }
+    }
+
+    /// Looks up transaction metadata by its transaction ID, returning
+    /// `Some(IncludedTransactionMetadata)` if found, or `None` if not found.
+    pub(super) fn get_by_tx_id(
+        &self,
+        tx_id: &TransactionId,
+    ) -> Option<&IncludedTransactionMetadata> {
+        self.transactions.get(tx_id)
+    }
+
+    /// Adds new transaction data to the cache. Cleans stale transactions before attempting to add
+    /// the new transaction. If the cache is full, removes the oldest transaction until the cache
+    /// is not full. Returns an error if the transaction ID already exists in the cache.
+    pub(super) fn add_transaction(
+        &mut self,
+        tx_id: TransactionId,
+        height: u64,
+        result: ExecTxResult,
```

Comment:
```suggestion
    pub(super) fn add(
        execution_results: HashMap<TransactionId, ExecTxResult>,
        block_height: u64,
```
The `_transaction` suffix doesn't really provide any more clarity imo, and is probably even a bit misleading since we're not actually adding a tx.

Also, I think we should handle the data in bulk for a couple of reasons: it avoids multiple calls to `clean_stale`, and it also avoids multiple calls to `Instant::now()`.  I'd even be inclined to have `clean_stale` take an `Instant` as an arg so it can avoid the syscall to the clock when it's called from inside this method.